### PR TITLE
fix(timestamp): fix some timestamp issue in watermark derivation

### PR DESCRIPTION
**Pull request by soundOfDestiny**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

fix some timestamp issue in watermark derivation due to daylight saving time

## Checklist For Contributors

- [x] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: src/frontend/src/expr/mod.rs [hunk 1]
```diff
@@ -772,7 +772,14 @@ impl ExprImpl {
                     ExprType::Add | ExprType::Subtract => {
                         let (_, lhs, rhs) = function_call.clone().decompose_as_binary();
                         if let ExprImpl::InputRef(input_ref) = &lhs && rhs.is_const() {
-                            Some((input_ref.index(), Some((expr_type, rhs))))
+                            if rhs.return_type() == DataType::Interval && rhs.as_literal().map_or(true, |literal| literal.get_data().as_ref().map_or(false, |scalar| {
```
Comment (xiangjinwu) [inline src/frontend/src/expr/mod.rs:775]: Checking my understanding:
* The default `true` after `as_literal` means we return `None` for non-literal
* The default `false` after `get_data` means we return `Some` when it is literal `null`

Maybe add some comments or tests to help make this clear.

BTW:
* The closed PR #9647 updates `is_now_offset`, and here we update `as_input_offset`. How are they related? Do we need to fix both?
* `is_const` then `as_literal` misses the case when rhs is `interval + interval` or `interval * number`. Practically it should be rare. Just making the difference clear.

File: src/frontend/src/expr/utils.rs [hunk 2]
```diff
@@ -498,15 +498,32 @@ impl WatermarkAnalyzer {
...
-            | ExprType::AtTimeZone => match self.visit_binary_op(func_call.inputs()) {
+            | ExprType::AtTimeZone) => match self.visit_binary_op(func_call.inputs()) {
                 (WatermarkDerivation::Constant, WatermarkDerivation::Constant) => {
                     WatermarkDerivation::Constant
                 }
                 (WatermarkDerivation::Watermark(idx), WatermarkDerivation::Constant) => {
-                    WatermarkDerivation::Watermark(idx)
+                    if ty == ExprType::AtTimeZone
+                        && !(func_call.return_type() == DataType::Timestamptz
+                            && func_call.inputs()[0].return_type() == DataType::Timestamp)
```
Comment (xiangjinwu) [inline src/frontend/src/expr/utils.rs:511]: What about writing the case `func_call.return_type() == DataType::Timestamp && func_call.inputs()[0].return_type() == DataType::Timestamptz` directly without `!`?

File: src/frontend/src/expr/utils.rs [hunk 3]
```diff
@@ -498,15 +498,32 @@ impl WatermarkAnalyzer {
...
+                        && func_call.inputs()[1]
+                            .as_literal()
+                            .and_then(|literal| literal.get_data().as_ref())
+                            .map_or(true, |time_zone| {
+                                if let ScalarImpl::Utf8(time_zone_str) = time_zone {
+                                    !time_zone_str.eq_ignore_ascii_case("UTC")
+                                } else {
+                                    true
```
Comment (xiangjinwu) [inline src/frontend/src/expr/utils.rs:519]: Maybe unreachable and we should panic here? Just like `as_interval()` used above.

File: src/frontend/src/expr/utils.rs (hunk 2 above)
Comment (soundOfDestiny) [reply src/frontend/src/expr/utils.rs:511]: > What about writing the case `func_call.return_type() == DataType::Timestamp && func_call.inputs()[0].return_type() == DataType::Timestamptz` directly without `!`?

I was afraid that in the future we might support conversion from one timestamp to another timestamp.

File: src/frontend/src/expr/mod.rs (hunk 1 above)
Comment (soundOfDestiny) [reply src/frontend/src/expr/mod.rs:775]: > Checking my understanding:
> 
> * The default `true` after `as_literal` means we return `None` for non-literal
> * The default `false` after `get_data` means we return `Some` when it is literal `null`
> 
> Maybe add some comments or tests to help make this clear.
> 
> BTW:
> 
> * The closed PR [feat(now): streaming now() with a timezone #9647](https://github.com/risingwavelabs/risingwave/pull/9647) updates `is_now_offset`, and here we update `as_input_offset`. How are they related? Do we need to fix both?
> * is_const then as_literal misses the case when rhs is interval + interval or interval * number. Practically it should be rare. Just making the difference clear.
Yes, it's the same as your understanding.
I have just added some explanation, thanks.

> How are they related? Do we need to fix both?
IIUC, even if we do not have monotonicity, we can still handle `now`, although we cannot do state cleaning in that case. Therefore we do not need to fix `is_now_offset`.
On the other hand, `as_input_offset` involves in some calculation on watermark so we need to fix this.

> is_const then as_literal misses the case when rhs is interval + interval or interval * number. Practically it should be rare. Just making the difference clear.

I've added an example.

File: src/frontend/src/expr/utils.rs (hunk 3 above)
Comment (soundOfDestiny) [reply src/frontend/src/expr/utils.rs:519]: > Maybe unreachable and we should panic here? Just like `as_interval()` used above.

fixed.

**Review by xiangjinwu — APPROVED** [review]

**Pull request merged**